pkg/neotest/signer.go

```diff
@@ -179,3 +188,57 @@ func checkMultiSigner(t testing.TB, s Signer) {
 		require.Equal(t, h, accs[i].Contract.ScriptHash(), "inconsistent multi-signer accounts")
 	}
 }
+
+type contractSigner struct {
+	params     func(tx *transaction.Transaction) []any
+	scriptHash util.Uint160
+}
+
+// NewContractSigner returns a contract signer for the provided contract hash.
+// getInvParams must return params to be used as invocation script for contract-based witness.
+func NewContractSigner(h util.Uint160, getInvParams func(tx *transaction.Transaction) []any) ContractSigner {
+	return &contractSigner{
+		scriptHash: h,
+		params:     getInvParams,
+	}
+}
+
+// InvocationScript implements ContractSigner.
+func (s *contractSigner) InvocationScript(tx *transaction.Transaction) ([]byte, error) {
+	params := s.params(tx)
+	script := io.NewBufBinWriter()
+	for i := range params {
+		emit.Any(script.BinWriter, params[i])
+	}
+	if script.Err != nil {
+		return nil, script.Err
+	}
+	return script.Bytes(), nil
+}
+
+// Script implements ContractSigner.
+func (s *contractSigner) Script() []byte {
+	return nil
+}
+
+// ScriptHash implements ContractSigner.
+func (s *contractSigner) ScriptHash() util.Uint160 {
+	return s.scriptHash
+}
+
+// SignHashable implements ContractSigner.
+func (s *contractSigner) SignHashable(uint32, hash.Hashable) []byte {
+	return nil
+}
+
+// SignTx implements ContractSigner.
+func (s *contractSigner) SignTx(magic netmode.Magic, tx *transaction.Transaction) error {
+	sc, err := s.InvocationScript(tx)
+	if err != nil {
+		return err
+	}
+	tx.Scripts = append(tx.Scripts, transaction.Witness{
+		InvocationScript: sc,
+	})
```

Comment:
We need to iterate over transaction signers and find the signer that matches the provided contract to put witness to the proper place.